### PR TITLE
adapter: teach plan insights about fast path clusters

### LINKED ISSUE #20804: Notice when another cluster has some indexes that would help
### Feature request

A common mistake is to create an index on a different cluster from the cluster that runs the queries that would need the index. We could catch these situations as follows: in addition to the normal optimization, we would run the optimizer with all indexes from all the clusters, and see whether more indexes would be used that way. If yes, then we can point out those indexes in a notice, with also printing in which cluster they are on.

To reduce the overhead, we should only do this if

1. the normal optimization yielded a non-fast path query, and
2. the normal optimization completed quickly.

Edit: In the meantime, I'm thinking about an alternative way to do this: `prune_and_annotate_dataflow_index_imports` should also gather situations when the dataflow could use an index that is not available. I think the only tricky part is that sometimes we have to jump through an MFP that JoinImplementation would lift away if an index were available. Note that this would be useful also for several other issues in the guardrails epic.

### PR DESCRIPTION
**Pull request by maddyblue**
For slow path queries in a plan insights context, try optimizing in all other user clusters and list the clusters that would have resulted in a fast path plan.

Fixes #20804

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/adapter/src/explain/insights.rs [hunk 1]
```diff
@@ -28,6 +56,45 @@ pub struct PlanInsights {
...
+                Ok(global_lir_plan)
+            })();
+            let Ok(plan) = res else {
+                continue;
+            };
+            let (plan, _, _) = plan.unapply();
+            if matches!(plan, PeekPlan::FastPath(_)) {
+                self.fast_path_clusters.insert(name);
+            }
+        }
```
Comment (jkosh44) [inline src/adapter/src/explain/insights.rs:96]: It might be worth optimizing in parallel since sometimes optimization can take awhile.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/explain/insights.rs (hunk 1 above)
Comment (maddyblue) [reply src/adapter/src/explain/insights.rs]: Done. This made a bunch of stuff async but that seems fine.

**Comment by ggevay** [comment]
I'm a bit concerned about making as many optimizer calls as there are clusters. E.g., https://github.com/MaterializeInc/accounts/issues/3 has 15 clusters.

How about doing all this only when the query is simple enough that there is a chance of fast path. `create_fast_path_plan` shows when a query might become a fast path query in general. But you don't need to do the entire machinery from there: I think the only case that is relevant here is checking whether the entire MIR expression of the query is just an MFP on top of a Get. You could check for this by calling `MapFilterProject::extract_from_expression` and then checking that the returned expression is a `MirRelationExpr::Get` of a `GlobalId`. Alternatively, you could simply call `MirRelationExpr::size()`, and check that it's smaller than, say, 10.

**Comment by maddyblue** [comment]
This only runs when

1. the plan insights notice is enabled (which only the console does)
2. the original optimization took < 10ms

Does that make you feel better? Also it's done in parallel. So even in the worst case (many clusters at 9.9ms per), it's not going to add more than 80ms in the 15 cluster scenario (assuming 2 cores for envd: 10ms * 8 clusters). And since it'll only happen in the console, I think it's worth it.

**Comment by ggevay** [comment]
Ok, sounds good, these safeguards seem enough!

File: src/adapter/src/coord/sequencer/inner/peek.rs [hunk 2]
```diff
@@ -609,6 +620,30 @@ impl Coordinator {
...
+                        let mut needs_plan_insights = explain_ctx.needs_plan_insights();
+                        let opt_limit = mz_adapter_types::dyncfgs::PLAN_INSIGHTS_NOTICE_FAST_PATH_CLUSTERS_OPTIMIZE_DURATION.get(catalog.system_config().dyncfgs());
+                        // However, disable this if we only want the notice and plan optimization
+                        // took longer than the threshold. This is to prevent a situation where
+                        // optimizing takes a while and there a lots of clusters, which would delay
+                        // peek execution by the product of those.
+                        if matches!(explain_ctx, ExplainContext::PlanInsightsNotice(_))
+                            && optimizer.duration() > opt_limit
+                        {
+                            needs_plan_insights = false;
```
Comment (ggevay) [inline src/adapter/src/coord/sequencer/inner/peek.rs:633]: This disables the entire plan insights machinery? I thought only the `compute_fast_path_clusters` insight would need to be disabled.

Comment (maddyblue) [reply src/adapter/src/coord/sequencer/inner/peek.rs:633]: Good catch! I'll plumb this down so it only restricts the slow parts.

**Pull request merged**